[18.0][MIG] hr_expense_invoice: Migration to version 18.0

david-banon-tecnativa opened the pull request:
@tecnativa @pedrobaeza @victoralmau 
Superseedes https://github.com/OCA/hr-expense/pull/282 by @cav-adhoc while maintaining commit ownership.

I've made some minor adjustments, orignal PR looked good to me and only needed a rebase to work.

TT55529

pedrobaeza commented:
/ocabot migration hr_expense_invoice

pedrobaeza commented on hr_expense_invoice/models/hr_expense_sheet.py:
```diff
@@ -180,13 +178,13 @@ def _compute_state(self):
         expense_sheets_with_invoices = self.get_expense_sheets_with_invoices()
         own_account_sheets = self.filtered(
             lambda sheet: sheet.payment_mode == "own_account"
+            and not sheet.advance_sheet_id
```
This can't be put here, as it belongs to other module AFAIK.

rafaelbn commented:
Hello,

- Create some expenses, couple of them invoices couple of them not.
- Create with this 4 expenses an expense report
- In the lines which are Invoices clic "Create Vendor Bill" (draft invoices will be created)
- In the Expense report summary clic in button "Approve" in the top-left corner and you get this error
- Video test: https://www.loom.com/share/a5bf3ed6f2214356b31531f64149f4b4?sid=ae626667-9cda-4e10-8999-07e51e797f30


```
> Odoo Server Error
> RPC_ERROR
> 
> Odoo Server Error
> 
> Occured on oca-hr-expense-18-0-pr308-b683208461fa.runboat.odoo-community.org on model hr.expense.sheet on 2025-08-04 22:37:26 GMT
> 
> Traceback (most recent call last):
>   File "/opt/odoo/odoo/http.py", line 2123, in _transactioning
>     return service_model.retrying(func, env=self.env)
>   File "/opt/odoo/odoo/service/model.py", line 156, in retrying
>     result = func()
>   File "/opt/odoo/odoo/http.py", line 2090, in _serve_ir_http
>     response = self.dispatcher.dispatch(rule.endpoint, args)
>   File "/opt/odoo/odoo/http.py", line 2338, in dispatch
>     result = self.request.registry['ir.http']._dispatch(endpoint)
>   File "/opt/odoo/odoo/addons/base/models/ir_http.py", line 333, in _dispatch
>     result = endpoint(**request.params)
>   File "/opt/odoo/odoo/http.py", line 754, in route_wrapper
>     result = endpoint(self, *args, **params_ok)
>   File "/opt/odoo/addons/web/controllers/dataset.py", line 42, in call_button
>     action = call_kw(request.env[model], method, args, kwargs)
>   File "/opt/odoo/odoo/api.py", line 535, in call_kw
>     result = getattr(recs, name)(*args, **kwargs)
>   File "/opt/odoo/addons/hr_expense/models/hr_expense_sheet.py", line 595, in action_approve_expense_sheets
>     self._do_approve()
>   File "/mnt/data/odoo-addons-dir/hr_expense_invoice/models/hr_expense_sheet.py", line 185, in _do_approve
>     own_account_sheets = self.filtered(
>   File "/opt/odoo/odoo/models.py", line 6523, in filtered
>     return self.browse(rec.id for rec in self if func(rec))
>   File "/opt/odoo/odoo/models.py", line 6222, in browse
>     ids = tuple(ids)
>   File "/opt/odoo/odoo/models.py", line 6523, in <genexpr>
>     return self.browse(rec.id for rec in self if func(rec))
>   File "/mnt/data/odoo-addons-dir/hr_expense_invoice/models/hr_expense_sheet.py", line 187, in <lambda>
>     and not sheet.advance_sheet_id
> AttributeError: 'hr.expense.sheet' object has no attribute 'advance_sheet_id'
> 
> The above server error caused the following client error:
> RPC_ERROR: Odoo Server Error
>     RPCError@http://oca-hr-expense-18-0-pr308-b683208461fa.runboat.odoo-community.org/web/assets/3894c65/web.assets_web.min.js:3145:338
>     makeErrorFromResponse@http://oca-hr-expense-18-0-pr308-b683208461fa.runboat.odoo-community.org/web/assets/3894c65/web.assets_web.min.js:3148:163
>     rpc._rpc/promise</<@http://oca-hr-expense-18-0-pr308-b683208461fa.runboat.odoo-community.org/web/assets/3894c65/web.assets_web.min.js:3153:34
>     
```

Le me know please, when I can test again. Thank you!

rafaelbn commented:
@david-banon-tecnativa Is this PR going to be worked on? Thank you!

david-banon-tecnativa commented:
Sorry for the delay @rafaelbn, got sidetracked with other tasks, I just pushed a working version, with the fix for #302 included. But now tests for `hr_expense_advance_clearing` are failing once again.

victoralmau commented on hr_expense_invoice/models/hr_expense_sheet.py:
```diff
@@ -0,0 +1,271 @@
+
+from odoo import Command, api, fields, models
+from odoo.exceptions import RedirectWarning, UserError
+from odoo.tools import float_compare
```
```suggestion
from odoo.tools import config, float_compare
```

victoralmau commented on hr_expense_invoice/models/hr_expense_sheet.py:
```diff
@@ -0,0 +1,271 @@
+            Command.create(expense._prepare_move_lines_vals())
+            for expense in expenses_without_invoice
+        ]
+
```
```suggestion
        test_condition = not config["test_enable"] or self.env.context.get(
            "test_hr_expense_invoice"
        )
        if test_condition:
            expenses_without_invoice = self.expense_line_ids.filtered(
                lambda r: not r.invoice_id
            )
            res["line_ids"] = [
                Command.create(expense._prepare_move_lines_vals())
                for expense in expenses_without_invoice
            ]
```

victoralmau commented on hr_expense_invoice/tests/test_hr_expense_invoice.py:
```diff
@@ -0,0 +1,260 @@
+class TestHrExpenseInvoice(TestExpenseCommon):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
```
```suggestion
        super().setUpClass()
        cls.env = cls.env(context=dict(cls.env.context, test_hr_expense_invoice=True))
```
I believe this will resolve the testing issue, similar to what happens in https://github.com/OCA/hr-expense/pull/307 (`hr_expense_invoice` + `hr_expense_advance_clearing` change the `_prepare_bills_vals()` method, so it is “_normal_” for the tests to be incorrect).

Andrii9090 reviewed (approved):
Funcional and code review :+1:
Without module installed
![hr_expense_invoice_4](https://github.com/user-attachments/assets/3f286b00-d3fb-4910-ac45-7195395e0549)

With this module we can `Create Vendor Bill` for each line. New invoices linked with selected expense lines.
![hr_expense_invoice_1](https://github.com/user-attachments/assets/00f825db-37df-482f-8f92-8b6762d3aae9)
![hr_expense_invoice_2](https://github.com/user-attachments/assets/c50d01cd-d39e-4d66-983a-75e7d37c627a)

BhaveshHeliconia reviewed (approved):
Functional review LGTM!

JordiBForgeFlow commented:
@Gelojr I cannot reproduce your error. 

When I create a vendor bill from the expense report and is in draft:

<img width="1451" height="737" alt="image" src="https://github.com/user-attachments/assets/db48cd94-c0c4-4882-8118-fdb5d0f6bc83" />

Then, before posting the bill I go to the expense report and try to Post the journal entries I get the right error message:

<img width="1498" height="641" alt="image" src="https://github.com/user-attachments/assets/f85bd88c-cb9b-4d9e-bdde-05682b5ce548" />

Gelojr commented:
Hi Jordi,
Thanks for double-checking. On my side, when the vendor bill is still in
Draft I do get the expected warning when trying to post the expense, so
that path looks correct. In the attached expense, however, the Vendor Bill
field shows “Draft bill” while the expense ended up posted.

My recollection is that I clicked “Create Vendor Bill” and then clicked
“Post” on the expense without posting the bill first, and it auto-confirmed
at that moment. I can’t confirm it 100% and I don’t have reproducible
steps, but that’s what I remember from that record. If your current checks
are working as expected, I’m fine with that—sharing the expense just for
reference.
http://oca-hr-expense-18-0-pr308-5218e49283a0.runboat.odoo-community.org/odoo/expense-reports/14

El lun, 25 ago 2025 a las 16:11, Jordi Ballester Alomar (<
***@***.***>) escribió:

> *JordiBForgeFlow* left a comment (OCA/hr-expense#308)
> <https://github.com/OCA/hr-expense/pull/308#issuecomment-3220466564>
>
> @Gelojr <https://github.com/Gelojr> I cannot reproduce your error.
>
> When I create a vendor bill from the expense report and is in draft:
> image.png (view on web)
> <https://github.com/user-attachments/assets/db48cd94-c0c4-4882-8118-fdb5d0f6bc83>
>
> Then, before posting the bill I go to the expense report and try to Post
> the journal entries I get the right error message:
> image.png (view on web)
> <https://github.com/user-attachments/assets/f85bd88c-cb9b-4d9e-bdde-05682b5ce548>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/OCA/hr-expense/pull/308#issuecomment-3220466564>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BDFUBLRSFKU7RGD3IFHLAPD3PMKSRAVCNFSM6AAAAACAWJH6PGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTEMRQGQ3DMNJWGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

JordiBForgeFlow commented:
@Gelojr I can reproduce the issue you mention when I create an expense report created with two expenses, where I only create a related invoice for one of them. Then I was able to post the expense report while keeping one of the vendor bill in draft.

This is not an issue with the migration of the module, as this bug exists also in prior versions. The problem is here: https://github.com/OCA/hr-expense/blob/17.0/hr_expense_invoice/models/hr_expense_sheet.py#L21 and https://github.com/OCA/hr-expense/pull/308/files#diff-406b9f2af65ef4b7b24c6638a958cf7301ddad72e846bc1c68083c281238282dR15

The method 
   ```
 def get_expense_sheets_with_invoices(self):
        return self.filtered(
            lambda sheet: all(expense.invoice_id for expense in sheet.expense_line_ids)
        )
```
Sould be using `any` not `all`

@david-banon-tecnativa do you agree? In any case we should handle this in a separate PR/Issue.

Gelojr commented:
Awesome, Jordi,
congrats on reproducing it! I didn’t remember the exact steps; I only
recall I managed to trigger it once. Great that you’ve identified it so it
can be fixed.

El lun, 25 ago 2025 a las 17:33, Jordi Ballester Alomar (<
***@***.***>) escribió:

> *JordiBForgeFlow* left a comment (OCA/hr-expense#308)
> <https://github.com/OCA/hr-expense/pull/308#issuecomment-3220754052>
>
> @Gelojr <https://github.com/Gelojr> I can reproduce the issue you mention
> when I create an expense report created with two expenses, where I only
> create a related invoice for one of them. Then I was able to post the
> expense report while keeping one of the vendor bill in draft.
>
> This is not an issue with the migration of the module, as this bug exists
> also in prior versions. The problem is here:
> https://github.com/OCA/hr-expense/blob/17.0/hr_expense_invoice/models/hr_expense_sheet.py#L21
> .
>
> The method
>
> def get_expense_sheets_with_invoices(self):
>      return self.filtered(
>          lambda sheet: all(expense.invoice_id for expense in sheet.expense_line_ids)
>      )
>
> Sould be using any not all
>
> @david-banon-tecnativa <https://github.com/david-banon-tecnativa> do you
> agree? In any case we should handle this in a separate PR/Issue.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/OCA/hr-expense/pull/308#issuecomment-3220754052>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BDFUBLVIK4LJSRIFSCUJCNT3PMUD3AVCNFSM6AAAAACAWJH6PGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTEMRQG42TIMBVGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

pedrobaeza commented:
Let's merge for now this one:

/ocabot merge nobump

pedrobaeza commented:
@david-banon-tecnativa please rebase and check the failing CI. You can at the same time add a commit with the fix for the other problem.

rafaelbn commented:
Hello @pedrobaeza ,

Is it a good idea merging this one with and existing bug?

> This is not an issue with the migration of the module, as this bug exists also in prior versions

Fixing the bug is not a choice? **Do you have an idea how do you want to FIX it?**

Thank you!
Regards

PD @Gelojr @ArantxaSudon  take care so in: 

> Known issues / Roadmap]
> - Multiple payment terms for a supplier invoice are not handled correctly.
> - Partial reconcile supplier invoices are also not correctly handled.

pedrobaeza commented:
Yes, the migration is correct as the problem is the same for all the versions, so you get at least the same module as in 17.0. Rest of the issues can be tackled in any moment. As now there's other problem, the fix may be included here, but you shouldn't block a migration because the discovering of a bug not related to the migration itself.

david-banon-tecnativa commented:
I’ve updated the tests and implemented a fix for expense sheets that include one sheet with an unposted invoice and another with no invoice, ensuring an error is raised as it shouldn't be postable until the invoice is confirmed.
I will backport to earlier versions once merged.

pedrobaeza commented:
@rafaelbn @Gelojr update your reviews.

pedrobaeza commented:
No need to bump version now for pip availability:

/ocabot merge nobump
